src/main/java/gov/cms/madie/terminology/service/FhirTerminologyService.java

```diff
@@ -278,4 +277,104 @@ private void updateOrInsertAllCodeSystems(List<CodeSystem> codeSystemList) {
       }
     }
   }
+
+  public List<Code> retrieveCodesAndCodeSystems(List<Map<String, String>> codeList, String apiKey) {
+    return codeList.stream()
+        .map(
+            code -> {
+              List<CodeSystemEntry> codeSystemEntries = mappingService.getCodeSystemEntries();
+              String codeName = code.get("code");
+              String codeSystemName = code.get("codeSystem");
+              String oid = code.get("oid") != null ? code.get("oid").replaceAll("'|'", "") : null;
+
+              Optional<Map.Entry<String, String>> mappedVersion =
+                  mapToFhirVersion(code.get("version"), oid, codeSystemEntries);
+
+              if (mappedVersion.isPresent()) {
+                String vsacVersion = mappedVersion.get().getKey();
+                String fhirVersion = mappedVersion.get().getValue();
+
+                return retrieveCodes(
+                    codeName, codeSystemName, vsacVersion, fhirVersion, oid, apiKey);
+              }
+
+              return retrieveCodes(codeName, codeSystemName, null, null, oid, apiKey);
+            })
+        .collect(Collectors.toList());
+  }
+
+  private Optional<Map.Entry<String, String>> mapToFhirVersion(
+      String version, String oid, List<CodeSystemEntry> codeSystemEntries) {
+
+    if (oid == null) {
+      return Optional.empty();
+    }
+
+    Optional<Map.Entry<String, String>> result;
+    if (version == null) {
+      result =
+          codeSystemEntries.stream()
+              .filter(codeSystemEntry -> codeSystemEntry.getOid().toString().equals(oid))
+              .findFirst()
+              .map(
+                  codeSystemVersion ->
+                      Map.entry(
+                          codeSystemVersion.getVersions().get(0).getVsac(),
+                          codeSystemVersion.getVersions().get(0).getFhir()));
+    } else {
+      result =
+          codeSystemEntries.stream()
+              .filter(codeSystemEntry -> codeSystemEntry.getOid().toString().equals(oid))
+              .flatMap(codeSystemEntry -> codeSystemEntry.getVersions().stream())
+              .filter(codeSystemVersion -> codeSystemVersion.getVsac().equals(version))
+              .map(
+                  codeSystemVersion ->
+                      Map.entry(codeSystemVersion.getVsac(), codeSystemVersion.getFhir()))
+              .findFirst();
+    }
+
+    if (result.isPresent()) {
+      return result;
+    }
+    return result;
```

Comment:
```suggestion
    return result;
```
Can be simplified to the above.

Reply:
thank you, I just fixed it